### PR TITLE
check if position_ids exists before using it

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @ArthurZucker  and @younesbelkada 

I think we should check if `position_ids` is `None` or not before we use it.

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -1274,7 +1274,11 @@ def prepare_inputs_for_generation(
 
         # TODO @gante we should only keep a `cache_position` in generate, and do +=1.
         # same goes for position ids. Could also help with continued generation.
-        cache_position = torch.arange(past_length, past_length + position_ids.shape[-1], device=position_ids.device)
+        cache_position = (
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1277]: the attention mask is always passed to the model so the position ids are always created before this!

Comment (jiqing-feng) [reply src/transformers/models/llama/modeling_llama.py:1277]: If so, should we set `attention_mask` as a required parameter? like 
```python
def prepare_inputs_for_generation(
        self, input_ids, attention_mask, past_key_values=None, inputs_embeds=None, **kwargs
    ):
```

Because we cannot control user behavior, I think we should avoid this error in codes or set a warning.

Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:1277]: have not seen any issue so far. If past key values is not None, the attention mask is created as well and always passed to the model by the generate function. Let's just check if positions ids exist or just use input_ids instead for device placement and shape as it always exists

Comment (jiqing-feng) [reply src/transformers/models/llama/modeling_llama.py]: Great idea! Have fixed it according to your comments:)

**Comment by jiqing-feng** [comment]
The CI is weird, I can pass the CI locally, could you please help to re-run the CI? Thx!

**Review by ArthurZucker — APPROVED** [review]
LGTM let's have a second look from @gante

**Comment by ArthurZucker** [comment]
feel free to merge from main and run `make fixup` to make sure CIs go green

**Comment by jiqing-feng** [comment]
The CI still failed, and it is my local CI result:
image

**Review by gante — APPROVED** [review]

**Comment by gante** [comment]
@jiqing-feng I hope you don't mind, I took the liberty to fix the failing test case 🤗 TL;DR when `inputs_embeds` is passed, we can't rely on `input_ids` to get the correct input length

**Comment by jiqing-feng** [comment]
> @jiqing-feng I hope you don't mind, I took the liberty to fix the failing test case 🤗 TL;DR when `inputs_embeds` is passed, we can't rely on `input_ids` to get the correct input length

Thanks for your fix!
